snuba/migrations/snuba_migrations/transactions/0002_transactions_onpremise_fix_orderby_and_partitionby.py

```diff
@@ -0,0 +1,117 @@
+    )
+
+    sampling_key_needs_update = curr_sampling_key != new_sampling_key
+    partition_key_needs_update = curr_partition_key != new_partition_key
+
+    if not sampling_key_needs_update and not partition_key_needs_update:
+        # Already up to date
+        return
+
+    # Create transactions_local_new and insert data
+    ((curr_create_table_statement,),) = clickhouse.execute(
+        f"SHOW CREATE TABLE {database}.{table_name}"
+    )
+
+    new_create_table_statement = curr_create_table_statement.replace(
+        table_name, table_name_new
+    )
+
+    # Insert sample clause before TTL
+    if sampling_key_needs_update:
+        assert "SAMPLE BY" not in new_create_table_statement
+        idx = new_create_table_statement.find("TTL")
+
+        new_create_table_statement = (
+            new_create_table_statement[:idx]
+            + f"SAMPLE BY {new_sampling_key} "
+            + new_create_table_statement[idx:]
+        )
+
+    # Switch the partition key
+    if partition_key_needs_update:
+        assert new_create_table_statement.count(curr_partition_key) == 1
+        new_create_table_statement = new_create_table_statement.replace(
+            curr_partition_key, new_partition_key
+        )
+
+    clickhouse.execute(new_create_table_statement)
+
+    # Copy over data in batches of 100,000
+    [(row_count,)] = clickhouse.execute(f"SELECT count(*) FROM {table_name}")
+    batch_size = 100000
+    batch_count = math.ceil(row_count / batch_size)
+
+    orderby = "toStartOfDay(finish_ts), project_id, event_id"
+
+    for i in range(batch_count):
+        skip = batch_size * i
+        clickhouse.execute(
+            f"""
+            INSERT INTO {table_name_new}
+            SELECT * FROM {table_name}
+            ORDER BY {orderby}
+            LIMIT {batch_size}
+            OFFSET {skip};
+            """
+        )
+
+    clickhouse.execute(f"DROP TABLE {table_name};")
+
+    clickhouse.execute(f"RENAME TABLE {table_name_new} to {table_name};")
+
+
+class Migration(migration.MultiStepMigration):
+    """
+    The first of two migrations that syncs the transactions_local table for onpremise
+    users migrating from versions of Snuba prior to the migration system.
+
+    This migration recreates the table if the sampling key is not present or the
+    partitioning key is not up to date. Since the sampling key was introduced later and
+    the partition key was changed, it's possible the user has a version of the table
+    where one or both of these is out of date. If this is the case we must recreate
+    the table and ensure data is copied over.
+    """
+
+    blocking = True  # This migration may take some time if there is data to migrate
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.RunPython(func=recreate_table),
+        ]
+
+    def backwards_local(self) -> Sequence[operations.Operation]:
+        return []
```

Comment:
Actually I think you can to a backwards migration if you did not swap the tables yet.

Reply:
There's not a backwards migration that should be done. Remember this just take you to the same state as you're actually already supposed to be in before this migration if you have a really old table.

Reply:
I might actually just delete the temp table in case it was created already here.

Reply:
If there are no backwards steps, the status will still be reverted right ?

Reply:
Also, if you drop the new table in the backwards step pay attention to the case where the original table has already been removed and the failure happened during the renaming. You would be without table at all.